src/main/java/mate/academy/EventManager.java

```diff
@@ -1,19 +1,64 @@
 package mate.academy;
 
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
 public class EventManager {
-    public void registerListener(EventListener listener) {
+    public static final int THREAD_POOL_MIN_SIZE = 1;
+    public static final int SHUTDOWN_TIMEOUT_MINUTES = 2;
+    private final ThreadPoolExecutor executor;
+    private final List<EventListener> listeners;
 
+    public EventManager() {
+        this.executor = new ThreadPoolExecutor(
+                THREAD_POOL_MIN_SIZE, THREAD_POOL_MIN_SIZE,
+                0L, TimeUnit.MILLISECONDS,
+                new LinkedBlockingQueue<>());
+        this.listeners = new CopyOnWriteArrayList<>();
     }
 
-    public void deregisterListener(EventListener listener) {
+    public void registerListener(EventListener listener) {
+        listeners.add(listener);
+        adjustThreadPoolSize();
+    }
 
+    public void deregisterListener(EventListener listener) {
+        listeners.remove(listener);
+        adjustThreadPoolSize();
     }
 
     public void notifyEvent(Event event) {
-
+        Iterator<EventListener> iterator = listeners.iterator();
+        while (iterator.hasNext()) {
+            EventListener listener = iterator.next();
+            executor.submit(() -> listener.onEvent(event));
+        }
     }
 
     public void shutdown() {
+        executor.shutdown();
+        try {
+            if (!executor.awaitTermination(SHUTDOWN_TIMEOUT_MINUTES, TimeUnit.MINUTES)) {
+                executor.shutdownNow();
+            }
+        } catch (InterruptedException e) {
+            executor.shutdownNow();
+            Thread.currentThread().interrupt();
+        }
+    }
```

Comment:
this logic that you have inside this method awaits 2 minutes(or end of tasks execution) and then immediately shuts down, returning the tasks that were still in queue if any, but we are ignoring them. I do not think that is what we want to do here since our method is called `shutdown()` and not `shutdownNow()`. So let's just call `executor.shutdown()`, maybe in a try/catch block, but nothing else, to make it gracefully, as our method name suggests

Reply:
It was ai-mentor's idea. Turn off the executor by force if any threads are stuck waiting too long.